### ISSUE TITLE
Bug: Healthcheck failures

### ISSUE THREAD

**Issue by egyptianbman** [aws/aws-app-mesh-roadmap#333]
**Summary**
What are you observing that doesn't seem right?

We are seeing services in app mesh fail healthchecks every few hours. The tasks are not logging any failures but get terminated due to the healthcheck failures. We don't have any healthchecks beyond the AWS-configured healthcheck for the envoy container `curl -s http://localhost:9901/server_info | grep state | grep -q LIVE`. We've routed the envoy container log to cloudwatch and have not observed any errors before termination.

**Steps to Reproduce**
What are the steps you can take to reproduce this issue?

1. Run a service with App Mesh configured.
2. Wait a few hours
3. Observe tasks will cycle in the events tab, due to healthcheck failures.

**Are you currently working around this issue?**
How are you currently solving this problem?

Unfortunately, no. Our hands seem to be tied. I tried scaling services to 2 so that if one task gets cycled due to a healthcheck failure, the other would be able to take requests. Even with 2 tasks, both tasks were stopped when the healthcheck failed.

**Additional context**
Anything else we should know?

I've verified this is not an application failure. We're running two different services and both seem to be restarting around the same time. When the healthcheck fails, the tasks are stopped before new tasks are started which causes an interruption to request processing.

**Comment by efe-selcuk** [comment]
Hey there. Just to be completely sure, the health checks reported failing are the envoy container, the events are suggesting that and not a health check from the load balancer?

Are you using ECS EC2 or Fargate?

**Comment by egyptianbman** [comment]
Hello, yes. I don't have healthchecks defined for the other containers so since the error message "failed container health checks", I can only assume it's the envoy container.

We are running ECS EC2.

**Comment by rupertbg** [comment]
Which App Mesh Envoy Image are you running @egyptianbman ?

**Comment by egyptianbman** [comment]
Hey @rupertbg, the services are currently running `840364872350.dkr.ecr.us-east-1.amazonaws.com/aws-appmesh-envoy:v1.16.1.1-prod`.

**Comment by rupertbg** [comment]
Looks like public.ecr.aws/appmesh/aws-appmesh-envoy:v1.17.2.0-prod makes a big improvement to this issue, haven't been able to pinpoint the cause though.

**Comment by dastbe** [comment]
@egyptianbman I'd also be interested in what log level you're running at (you can increase the log level by setting ENVOY_LOG_LEVEL environment variable to debug or trace). Envoy tends to log very little at INFO so getting debug logs may reveal something. It'd also be useful to see your healthcheck configuration to get an idea of over how long a period it was unhealthy over.

This container healthcheck failing is basically saying that Envoy is not able to connect to the Envoy Management Service. The connection itself should be cycled every ~30-35 minutes, but outside of that should remain stable.

**Comment by egyptianbman** [comment]
Thanks for the tips. I'll update the envoy container, enable `DEBUG` logging and report back shortly.

Regarding the healthchecks, I use the defaults that the AWS console configures:
```json
            "healthCheck": {
                "retries": 3,
                "command": [
                    "CMD-SHELL",
                    "curl -s http://localhost:9901/server_info | grep state | grep -q LIVE"
                ],
                "timeout": 2,
                "interval": 5,
                "startPeriod": 10
            },
```

**Comment by egyptianbman** [comment]
Sorry for the delay in getting back to this. I **finally** tracked down the healthcheck failure. I have to say, I absolutely hate how AWS doesn't properly report healthcheck failures. It makes tracking down issues like this extremely difficult and time consuming. Since the service stops randomly every few hours, I more often than not am not able to view the old task as going to the url for an old task redirects back to the cluster with an error message like:
```
Failed to describe task <task-id> for cluster <cluster>
Task not found
```

Even when I was able to catch the task before it was (presumably) deleted, I'm not able to get any useful information out of it, mainly:
![image](https://user-images.githubusercontent.com/323600/122994537-3260cc80-d36e-11eb-9dd6-276083f899ff.png)

Finally, I decided to just script my own `watch` of the envoy container until it does to try to identify what the failure is. I knew it was the envoy container because it's the only container in this service that has a health check so the failure had to be coming from there.

For anyone else running into a similar issue trying to debug health checks, this ended up working well for me:
```console
while true; do
        date
        id=$(docker ps | grep '<container-id-or-alias>' | awk '{ print $1 }')
        [[ -z ${id} ]] && echo ${last} && break                                                    
        last=$(docker inspect --format "{{json .State }}" ${id})
        sleep 1
done
```
Be sure to replace `<container-id-or-alias>`. This should basically print a timestamp every second until the container is deleted, at which point it you'll get an output of the state at shutdown. A few times running it gave me a successful container at the time it was stopped/deleted but finally, I got the information I needed:
```json
{
    "Status": "exited",
    "Running": false,
    "Paused": false,
    "Restarting": false,
    "OOMKilled": false,
    "Dead": false,
    "Pid": 0,
    "ExitCode": 0,
    "Error": "",
    "StartedAt": "2021-06-22T19:35:30.917256209Z",
    "FinishedAt": "2021-06-22T20:19:29.391638972Z",
    "Health": {
        "Status": "unhealthy",
        "FailingStreak": 0,
        "Log": [
            {
                "Start": "2021-06-22T20:18:57.061418279Z",
                "End": "2021-06-22T20:18:59.063775199Z",
                "ExitCode": -1,
                "Output": "Health check exceeded timeout (2s)"
            },
            {
                "Start": "2021-06-22T20:19:04.080485186Z",
                "End": "2021-06-22T20:19:06.080660907Z",
                "ExitCode": -1,
                "Output": "Health check exceeded timeout (2s)"
            },
            {
                "Start": "2021-06-22T20:19:11.108211269Z",
                "End": "2021-06-22T20:19:13.108423151Z",
                "ExitCode": -1,
                "Output": "Health check exceeded timeout (2s)"
            },
            {
                "Start": "2021-06-22T20:19:18.150166242Z",
                "End": "2021-06-22T20:19:20.15405589Z",
                "ExitCode": -1,
                "Output": "Health check exceeded timeout (2s)"
            },
            {
                "Start": "2021-06-22T20:19:25.17749678Z",
                "End": "2021-06-22T20:19:25.887188114Z",
                "ExitCode": 0,
                "Output": ""
            }
        ]
    }
}
```

Knowing what I know now, I'll try increasing the timeout to 5 seconds to see if that leads to more stability.

**Comment by Y0Username** [comment]
Hey @egyptianbman,
These HC timeouts may indicate that Envoy process is unreachable or died. 
Are you still seeing this issue? If yes, what are the logs in the Envoy container when the HC starts failing?

**Comment by Y0Username** [comment]
Closing this for now. If you see this issue again, please feel free to re-open.

**Comment by shwetasahuit** [comment]
Hi @egyptianbman , Is this still an issue that you are facing, we need more details on your case to figure out if there is an issue at appmesh side and provide help to resolve it. Please reach out to us directly through Aws Support. We will be happy to assist you.

**Comment by maestre3d** [comment]
Maybe this won't fix your issue but could help others.

I managed to fix health check issues using `APPMESH_RESOURCE_ARN` environment variable -_in `envoy` container_- pointing to my AppMesh `Virtual Node` ARN. `APPMESH_VIRTUAL_NODE_NAME` and `APPMESH_RESOURCE_NAME` environment variables are not supported by Envoy 1.15+ versions (we're at 1.23 ATM). Tutorials like [this](https://www.appmeshworkshop.com/introduction/) are using this old version and could lead newcomers to deploy misconfigured envoy nodes.

Furthermore, there is an [official guide](https://docs.aws.amazon.com/app-mesh/latest/userguide/getting-started-ecs.html) by AWS following this bad configuration, leading to unhealthy nodes and thus to invalid deployments. It is using the latest Envoy version (1.23) while still using `APPMESH_VIRTUAL_NODE_NAME` environment variable.

In addition, I've attached the AWS-Managed [AWSAppMeshEnvoyAccess](https://us-east-1.console.aws.amazon.com/iam/home#/policies/arn:aws:iam::aws:policy/AWSAppMeshEnvoyAccess) to my **ECS Task Role** (not task execution role nor ecsInstance role for EC2 backends).

Finally, at the end I managed to properly deploy my service using CloudMap, ECS with EC2/ASG and AppMesh.
